test/interceptor_test.go

```diff
@@ -277,3 +277,270 @@ func (s) TestChainStreamClientInterceptor_ContextValuePropagation(t *testing.T)
+	}
+	if val != nil {
+		t.Fatalf("stream interceptor failed: %v", val)
+	}
+}
+
+// TestCallOnDialStreamClientInterceptor_ContextValuePropagation verifies that stream interceptors
+// specified as DialOptions or as CallOptions receive context values specified in the original call
+// as well as the ones specified by interceptors in the chain.
+func (s) TestCallOnDialStreamClientInterceptor_ContextValuePropagation(t *testing.T) {
+	errCh := testutils.NewChannel()
+	dialInt := func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		if got, ok := ctx.Value(parentCtxkey{}).(string); !ok || got != parentCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("dial interceptor got %q in context.Val, want %q", got, parentCtxVal))
+		}
+		if ctx.Value(baseInterceptorCtxKey{}) != nil {
+			errCh.SendContext(ctx, fmt.Errorf("first interceptor should not have %T in context", baseInterceptorCtxKey{}))
+		}
+		dialCtx := context.WithValue(ctx, baseInterceptorCtxKey{}, baseInterceptorCtxVal)
+		return streamer(dialCtx, desc, cc, method, opts...)
+	}
+
+	callInt := func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		if got, ok := ctx.Value(parentCtxkey{}).(string); !ok || got != parentCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("call interceptor got %q in context.Val, want %q", got, parentCtxVal))
+		}
+		if got, ok := ctx.Value(baseInterceptorCtxKey{}).(string); !ok || got != baseInterceptorCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("call interceptor got %q in context.Val, want %q", got, baseInterceptorCtxVal))
+		}
+		errCh.SendContext(ctx, nil)
+		return streamer(ctx, desc, cc, method, opts...)
+	}
+
+	// Start a stub server and use the first stream interceptor while creating
+	// a ClientConn to it.
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			if _, err := stream.Recv(); err != nil {
+				return err
+			}
+			return stream.Send(&testpb.StreamingOutputCallResponse{})
+		},
+	}
+	if err := ss.Start(nil, grpc.WithStreamInterceptor(dialInt)); err != nil {
+		t.Fatalf("Failed to start stub server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Use the second interceptor while invoking a streaming RPC.
+	if _, err := ss.Client.FullDuplexCall(context.WithValue(ctx, parentCtxkey{}, parentCtxVal), grpc.CallStreamClientInterceptor(callInt)); err != nil {
+		t.Fatalf("ss.Client.FullDuplexCall() failed: %v", err)
+	}
+	val, err := errCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("timeout when waiting for stream interceptor to be invoked: %v", err)
+	}
+	if val != nil {
+		t.Fatalf("stream interceptor failed: %v", val)
+	}
+}
```

Comment:
Could you also please add a test for the following additional cases:
- verify that the filtering of interceptors works as expected
- verify that interceptors added by call option interceptors get executed in the expected order.